In [131]:
import numpy as np
import pandas as pd

In [132]:
filename = 'sales_predictions.csv'
df = pd.read_csv(filename)
df.head(15) 

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0            FDA15        9.300          Low Fat         0.016047   
1            DRC01        5.920          Regular         0.019278   
2            FDN15       17.500          Low Fat         0.016760   
3            FDX07       19.200          Regular         0.000000   
4            NCD19        8.930          Low Fat         0.000000   
5            FDP36       10.395          Regular         0.000000   
6            FDO10       13.650          Regular         0.012741   
7            FDP10          NaN          Low Fat         0.127470   
8            FDH17       16.200          Regular         0.016687   
9            FDU28       19.200          Regular         0.094450   
10           FDY07       11.800          Low Fat         0.000000   
11           FDA03       18.500          Regular         0.045464   
12           FDX32       15.100          Regular         0.100014   
13           FDS46       17.600          Regular         0.047257   
14           FDF32       16.350          Low Fat         0.068024   

                Item_Type  Item_MRP Outlet_Identifier  \
0                   Dairy  249.8092            OUT049   
1             Soft Drinks   48.2692            OUT018   
2                    Meat  141.6180            OUT049   
3   Fruits and Vegetables  182.0950            OUT010   
4               Household   53.8614            OUT013   
5            Baking Goods   51.4008            OUT018   
6             Snack Foods   57.6588            OUT013   
7             Snack Foods  107.7622            OUT027   
8            Frozen Foods   96.9726            OUT045   
9            Frozen Foods  187.8214            OUT017   
10  Fruits and Vegetables   45.5402            OUT049   
11                  Dairy  144.1102            OUT046   
12  Fruits and Vegetables  145.4786            OUT049   
13            Snack Foods  119.6782            OUT046   
14  Fruits and Vegetables  196.4426            OUT013   

    Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                        1999      Medium               Tier 1   
1                        2009      Medium               Tier 3   
2                        1999      Medium               Tier 1   
3                        1998         NaN               Tier 3   
4                        1987        High               Tier 3   
5                        2009      Medium               Tier 3   
6                        1987        High               Tier 3   
7                        1985      Medium               Tier 3   
8                        2002         NaN               Tier 2   
9                        2007         NaN               Tier 2   
10                       1999      Medium               Tier 1   
11                       1997       Small               Tier 1   
12                       1999      Medium               Tier 1   
13                       1997       Small               Tier 1   
14                       1987        High               Tier 3   

          Outlet_Type  Item_Outlet_Sales  
0   Supermarket Type1          3735.1380  
1   Supermarket Type2           443.4228  
2   Supermarket Type1          2097.2700  
3       Grocery Store           732.3800  
4   Supermarket Type1           994.7052  
5   Supermarket Type2           556.6088  
6   Supermarket Type1           343.5528  
7   Supermarket Type3          4022.7636  
8   Supermarket Type1          1076.5986  
9   Supermarket Type1          4710.5350  
10  Supermarket Type1          1516.0266  
11  Supermarket Type1          2187.1530  
12  Supermarket Type1          1589.2646  
13  Supermarket Type1          2145.2076  
14  Supermarket Type1          1977.4260

In [133]:
df['Item_Fat_Content'].unique()

array(['Low Fat', 'Regular', 'low fat', 'LF', 'reg'], dtype=object)

In [134]:
#Hay que transformar low fat y LF a Low Fat.
#Hay que transformar reg a Regular.
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('low fat','Low Fat')
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('LF','Low Fat')
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('reg','Regular')
df['Item_Fat_Content'].unique()

array(['Low Fat', 'Regular'], dtype=object)

In [135]:
df.shape

(8523, 12)

In [136]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [137]:
df.duplicated().sum() #No hay duplicados

0

In [138]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Outlet_Size tiene bastantes valores NaN, sin embargo, exisen los Outlet_Identifier, que son la "ID" de los outlet.
De esta manera, podemos conocer el outlet size por medio del outlet identifier.

In [139]:
#NO AGREGAR LA MEDIA, MIRAR EL IDENTIFIER Y CON ESO SACAR EL WEIGHT.
df['Outlet_Identifier'].unique()

array(['OUT049', 'OUT018', 'OUT010', 'OUT013', 'OUT027', 'OUT045',
       'OUT017', 'OUT046', 'OUT035', 'OUT019'], dtype=object)

In [140]:
df['Outlet_Size'].unique()

array(['Medium', nan, 'High', 'Small'], dtype=object)

In [141]:
df["Outlet_Size"] = df.groupby('Outlet_Type')['Outlet_Size'].transform(lambda x: x.ffill())
#apply
#Outlet_Type
#Outlet size tiene relación con outlet type

In [142]:
df.head(15)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0            FDA15        9.300          Low Fat         0.016047   
1            DRC01        5.920          Regular         0.019278   
2            FDN15       17.500          Low Fat         0.016760   
3            FDX07       19.200          Regular         0.000000   
4            NCD19        8.930          Low Fat         0.000000   
5            FDP36       10.395          Regular         0.000000   
6            FDO10       13.650          Regular         0.012741   
7            FDP10          NaN          Low Fat         0.127470   
8            FDH17       16.200          Regular         0.016687   
9            FDU28       19.200          Regular         0.094450   
10           FDY07       11.800          Low Fat         0.000000   
11           FDA03       18.500          Regular         0.045464   
12           FDX32       15.100          Regular         0.100014   
13           FDS46       17.600          Regular         0.047257   
14           FDF32       16.350          Low Fat         0.068024   

                Item_Type  Item_MRP Outlet_Identifier  \
0                   Dairy  249.8092            OUT049   
1             Soft Drinks   48.2692            OUT018   
2                    Meat  141.6180            OUT049   
3   Fruits and Vegetables  182.0950            OUT010   
4               Household   53.8614            OUT013   
5            Baking Goods   51.4008            OUT018   
6             Snack Foods   57.6588            OUT013   
7             Snack Foods  107.7622            OUT027   
8            Frozen Foods   96.9726            OUT045   
9            Frozen Foods  187.8214            OUT017   
10  Fruits and Vegetables   45.5402            OUT049   
11                  Dairy  144.1102            OUT046   
12  Fruits and Vegetables  145.4786            OUT049   
13            Snack Foods  119.6782            OUT046   
14  Fruits and Vegetables  196.4426            OUT013   

    Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                        1999      Medium               Tier 1   
1                        2009      Medium               Tier 3   
2                        1999      Medium               Tier 1   
3                        1998         NaN               Tier 3   
4                        1987        High               Tier 3   
5                        2009      Medium               Tier 3   
6                        1987        High               Tier 3   
7                        1985      Medium               Tier 3   
8                        2002        High               Tier 2   
9                        2007        High               Tier 2   
10                       1999      Medium               Tier 1   
11                       1997       Small               Tier 1   
12                       1999      Medium               Tier 1   
13                       1997       Small               Tier 1   
14                       1987        High               Tier 3   

          Outlet_Type  Item_Outlet_Sales  
0   Supermarket Type1          3735.1380  
1   Supermarket Type2           443.4228  
2   Supermarket Type1          2097.2700  
3       Grocery Store           732.3800  
4   Supermarket Type1           994.7052  
5   Supermarket Type2           556.6088  
6   Supermarket Type1           343.5528  
7   Supermarket Type3          4022.7636  
8   Supermarket Type1          1076.5986  
9   Supermarket Type1          4710.5350  
10  Supermarket Type1          1516.0266  
11  Supermarket Type1          2187.1530  
12  Supermarket Type1          1589.2646  
13  Supermarket Type1          2145.2076  
14  Supermarket Type1          1977.4260

De la misma manera, podemos encontrar los valores faltantes en Item_Weight con su identificador Item_Identifier.

In [143]:
# df['Item_Weight'] = df.groupby('Item_Identifier')['Item_Weight'].transform(lambda x: x.fillna(method='ffill'))
df["Item_Weight"] = df.groupby('Item_Identifier')['Item_Weight'].transform(lambda x: x.ffill())

In [144]:
df.head(15)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0            FDA15        9.300          Low Fat         0.016047   
1            DRC01        5.920          Regular         0.019278   
2            FDN15       17.500          Low Fat         0.016760   
3            FDX07       19.200          Regular         0.000000   
4            NCD19        8.930          Low Fat         0.000000   
5            FDP36       10.395          Regular         0.000000   
6            FDO10       13.650          Regular         0.012741   
7            FDP10          NaN          Low Fat         0.127470   
8            FDH17       16.200          Regular         0.016687   
9            FDU28       19.200          Regular         0.094450   
10           FDY07       11.800          Low Fat         0.000000   
11           FDA03       18.500          Regular         0.045464   
12           FDX32       15.100          Regular         0.100014   
13           FDS46       17.600          Regular         0.047257   
14           FDF32       16.350          Low Fat         0.068024   

                Item_Type  Item_MRP Outlet_Identifier  \
0                   Dairy  249.8092            OUT049   
1             Soft Drinks   48.2692            OUT018   
2                    Meat  141.6180            OUT049   
3   Fruits and Vegetables  182.0950            OUT010   
4               Household   53.8614            OUT013   
5            Baking Goods   51.4008            OUT018   
6             Snack Foods   57.6588            OUT013   
7             Snack Foods  107.7622            OUT027   
8            Frozen Foods   96.9726            OUT045   
9            Frozen Foods  187.8214            OUT017   
10  Fruits and Vegetables   45.5402            OUT049   
11                  Dairy  144.1102            OUT046   
12  Fruits and Vegetables  145.4786            OUT049   
13            Snack Foods  119.6782            OUT046   
14  Fruits and Vegetables  196.4426            OUT013   

    Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                        1999      Medium               Tier 1   
1                        2009      Medium               Tier 3   
2                        1999      Medium               Tier 1   
3                        1998         NaN               Tier 3   
4                        1987        High               Tier 3   
5                        2009      Medium               Tier 3   
6                        1987        High               Tier 3   
7                        1985      Medium               Tier 3   
8                        2002        High               Tier 2   
9                        2007        High               Tier 2   
10                       1999      Medium               Tier 1   
11                       1997       Small               Tier 1   
12                       1999      Medium               Tier 1   
13                       1997       Small               Tier 1   
14                       1987        High               Tier 3   

          Outlet_Type  Item_Outlet_Sales  
0   Supermarket Type1          3735.1380  
1   Supermarket Type2           443.4228  
2   Supermarket Type1          2097.2700  
3       Grocery Store           732.3800  
4   Supermarket Type1           994.7052  
5   Supermarket Type2           556.6088  
6   Supermarket Type1           343.5528  
7   Supermarket Type3          4022.7636  
8   Supermarket Type1          1076.5986  
9   Supermarket Type1          4710.5350  
10  Supermarket Type1          1516.0266  
11  Supermarket Type1          2187.1530  
12  Supermarket Type1          1589.2646  
13  Supermarket Type1          2145.2076  
14  Supermarket Type1          1977.4260

In [145]:
df.isna().sum()

Item_Identifier                0
Item_Weight                  272
Item_Fat_Content               0
Item_Visibility                0
Item_Type                      0
Item_MRP                       0
Outlet_Identifier              0
Outlet_Establishment_Year      0
Outlet_Size                    1
Outlet_Location_Type           0
Outlet_Type                    0
Item_Outlet_Sales              0
dtype: int64

In [146]:
df[df['Outlet_Size'].isna()]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3           FDX07         19.2          Regular              0.0   

               Item_Type  Item_MRP Outlet_Identifier  \
3  Fruits and Vegetables   182.095            OUT010   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type    Outlet_Type  \
3                       1998         NaN               Tier 3  Grocery Store   

   Item_Outlet_Sales  
3             732.38

In [147]:
df1 = df.loc[df['Outlet_Identifier'] == 'OUT010',]
df1['Outlet_Size'].value_counts()

Small    554
Name: Outlet_Size, dtype: int64

Debido a que todos los valores del identificador OUT010 son Small, reemplazo el valor de Outlet_Size por "Small"

In [148]:
df.loc[3,'Outlet_Size'] = 'Small'
df.isna().sum()

Item_Identifier                0
Item_Weight                  272
Item_Fat_Content               0
Item_Visibility                0
Item_Type                      0
Item_MRP                       0
Outlet_Identifier              0
Outlet_Establishment_Year      0
Outlet_Size                    0
Outlet_Location_Type           0
Outlet_Type                    0
Item_Outlet_Sales              0
dtype: int64

Aún me quedan valores NaN en Item_Weight (272). Debido a que esas 272 filas representan solo un 3.2% del total de los datos, decido eliminarlas.

In [149]:
df.dropna(subset=['Item_Weight'], inplace=True)

In [150]:
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

Valor mínimo, valor máximo y media de cada columna numérica

In [151]:
#Mínimo
df.select_dtypes(include=[float, int]).min()

Item_Weight                     4.555
Item_Visibility                 0.000
Item_MRP                       31.490
Outlet_Establishment_Year    1985.000
Item_Outlet_Sales              33.290
dtype: float64

In [152]:
#Máximo
df.select_dtypes(include=[float, int]).max()

Item_Weight                     21.350000
Item_Visibility                  0.328391
Item_MRP                       266.888400
Outlet_Establishment_Year     2009.000000
Item_Outlet_Sales            13086.964800
dtype: float64

In [153]:
#Media
df.select_dtypes(include=[float, int]).mean() 

Item_Weight                    12.879107
Item_Visibility                 0.065828
Item_MRP                      141.154212
Outlet_Establishment_Year    1998.254878
Item_Outlet_Sales            2176.106901
dtype: float64